Test/NumberTheory/FastPower.test.cpp

```diff
@@ -0,0 +1,28 @@
+#define FAST_POWER_TEST
+#define CATCH_CONFIG_MAIN
+
+#include "../catch.hpp"
+#include "../../NumberTheory/FastPower.cpp"
+
+
+TEST_CASE("Base cases", "[fast_power]") {
+    //REQUIRE(fast_power(0,0) == undefined);
+    REQUIRE(fast_power(0,1) == 0);
+    REQUIRE(fast_power(1,0) == 1);
+    REQUIRE(fast_power(1,1) == 1);
+}
+
+TEST_CASE("Normal cases", "[fast_power]") {
+    REQUIRE(fast_power(2,2) == 4);
+    REQUIRE(fast_power(2,4) == 16);
+    REQUIRE(fast_power(3,4) == 81);
+    REQUIRE(fast_power(7,9) == 40353607);
+    REQUIRE(fast_power(15,10) == 576650390625);
+}
+
+TEST_CASE("Overflow cases", "[fast_power]") {
+    REQUIRE(fast_power(2,100) == 68719476736);
+    REQUIRE(fast_power(10,99) == 4440381706574496940U);
```

Comment:
I have tried returning the result modulo several prime numbers. The number ```1000000000007``` seems to be suitable, as a number bigger than it will result in ```fast_power(2,100) ``` being greater than ```2^100 mod (2^64-1)```.
```fast_power(2,100)``` now returns ```40955801449```.
However, cases like ```fast_power(33,33)``` still cause overflow.
What do you think?